File: src/plugins/builtins/LeakyReLU.ts
import type { BlockPlugin } from "../types.js";
import { getNum } from "./_helpers.js";

const LeakyReLU: BlockPlugin = {
  inputs: ["x"],
  showDepth: false,
  outputs: ["y"],

  params: [{ name: "negative_slope", type: "number", required: false }],
  inferShape(inputs, _params) {
    if (inputs.length === 0) throw new Error("LeakyReLU requires an input");
    return [inputs[0]];
  },
  paramCount: () => 0,
  codegen: {
    pytorch(block, inputVars, outputVars) {
      const slope = getNum(block.params, "negative_slope") ?? 0.01;
      return {
        init: `self.${block.id} = nn.LeakyReLU(${slope})`,
        forward: `${outputVars[0]} = self.${block.id}(${inputVars[0]})`,
      };
    },
    keras(block, inputVars, outputVars) {
      const slope = getNum(block.params, "negative_slope") ?? 0.01;
      return {
        forward: `${outputVars[0]} = keras.layers.LeakyReLU(alpha=${slope})(${inputVars[0]})`,
      };
    },
    candle(block, inputVars, outputVars) {
      const slope = getNum(block.params, "negative_slope") ?? 0.01;
      return {
        forward: `${outputVars[0]} = ${inputVars[0]}.leaky_relu(${slope})?;`,
      };
    },
  },
};

export default LeakyReLU;

File: mlmd-builtin-plugins/src/blocks/leaky_relu.rs
// ---------------------------------------------------------------------------
// LeakyReLU activation block.
//
// Port of src/plugins/builtins/LeakyReLU.ts
// ---------------------------------------------------------------------------

use std::collections::HashMap;

use mlmd_core::types::*;

use crate::helpers::get_num;

// ---------------------------------------------------------------------------
// BlockDef for shape inference
// ---------------------------------------------------------------------------

struct LeakyReLUBlockDef;

impl BlockDef for LeakyReLUBlockDef {
    fn name(&self) -> &str {
        "LeakyReLU"
    }

    fn params(&self) -> &[ParamSpec] {
        static PARAMS: std::sync::LazyLock<Vec<ParamSpec>> = std::sync::LazyLock::new(|| {
            vec![ParamSpec {
                name: "negative_slope".into(),
                param_type: ParamType::Number,
                required: false,
                default: None,
            }]
        });
        &PARAMS
    }

    fn infer_shape(
        &self,
        inputs: &[Shape],
        _params: &HashMap<String, ParamValue>,
    ) -> Result<Vec<Shape>, String> {
        if inputs.is_empty() {
            return Err("LeakyReLU requires an input".to_string());
        }
        Ok(vec![inputs[0].clone()])
    }

    fn param_count(
        &self,
        _inputs: &[Shape],
        _params: &HashMap<String, ParamValue>,
    ) -> Option<usize> {
        Some(0)
    }

    fn show_depth(&self) -> bool {
        false
    }
}

// ---------------------------------------------------------------------------
// Registration
// ---------------------------------------------------------------------------

pub fn register() {
    register_block(Box::new(LeakyReLUBlockDef));

    register_block_codegen("LeakyReLU", "pytorch", pytorch_codegen);
    register_block_codegen("LeakyReLU", "keras", keras_codegen);
    register_block_codegen("LeakyReLU", "candle", candle_codegen);
}

// ---------------------------------------------------------------------------
// Codegen functions
// ---------------------------------------------------------------------------

fn pytorch_codegen(
    block: &Block,
    input_vars: &[String],
    output_vars: &[String],
) -> BlockCodegenResult {
    let slope = get_num(&block.params, "negative_slope").unwrap_or(0.01);
    BlockCodegenResult {
        init: Some(CandleInitOrString::Plain(format!(
            "self.{} = nn.LeakyReLU({})",
            block.id, slope
        ))),
        forward: format!(
            "{} = self.{}({})",
            output_vars.first().map(|s| s.as_str()).unwrap_or("?"),
            block.id,
            input_vars.first().map(|s| s.as_str()).unwrap_or("?"),
        ),
    }
}

fn keras_codegen(
    block: &Block,
    input_vars: &[String],
    output_vars: &[String],
) -> BlockCodegenResult {
    let slope = get_num(&block.params, "negative_slope").unwrap_or(0.01);
    BlockCodegenResult {
        init: None,
        forward: format!(
            "{} = keras.layers.LeakyReLU(alpha={})({})",
            output_vars.first().map(|s| s.as_str()).unwrap_or("?"),
            slope,
            input_vars.first().map(|s| s.as_str()).unwrap_or("?"),
        ),
    }
}

fn candle_codegen(
    block: &Block,
    input_vars: &[String],
    output_vars: &[String],
) -> BlockCodegenResult {
    let slope = get_num(&block.params, "negative_slope").unwrap_or(0.01);
    BlockCodegenResult {
        init: None,
        forward: format!(
            "{} = {}.leaky_relu({})?;",
            output_vars.first().map(|s| s.as_str()).unwrap_or("?"),
            input_vars.first().map(|s| s.as_str()).unwrap_or("?"),
            slope,
        ),
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_leaky_relu_block_def() {
        let def = LeakyReLUBlockDef;
        assert_eq!(def.name(), "LeakyReLU");
        assert!(!def.show_depth());

        let shapes = def
            .infer_shape(&[vec![3, 224, 224]], &HashMap::new())
            .unwrap();
        assert_eq!(shapes, vec![vec![3, 224, 224]]);

        assert!(def.infer_shape(&[], &HashMap::new()).is_err());
        assert_eq!(def.param_count(&[], &HashMap::new()), Some(0));
    }
}
